tsdb/db_test.go

```diff
@@ -3489,3 +3490,222 @@ func TestDBPanicOnMmappingHeadChunk(t *testing.T) {
+	m4 := metadata.Metadata{Type: "counter", Unit: "unit_4", Help: "help_4"}
+	m5 := metadata.Metadata{Type: "counter", Unit: "unit_5", Help: "help_5"}
+	app = db.Appender(ctx)
+	app.AppendMetadata(0, s1, m1)
+	app.AppendMetadata(0, s4, m4)
+	app.AppendMetadata(0, s2, m5)
+	app.Commit()
+
+	// Read the WAL to see if the disk storage format is correct.
+	recs := readTestWAL(t, path.Join(db.Dir(), "wal"))
+	var gotMetadataBlocks [][]record.RefMetadata
+	for _, rec := range recs {
+		if mr, ok := rec.([]record.RefMetadata); ok {
+			gotMetadataBlocks = append(gotMetadataBlocks, mr)
+		}
+	}
+
+	expectedMetadata := []record.RefMetadata{
+		{Ref: 1, Type: record.GetMetricType(m1.Type), Unit: m1.Unit, Help: m1.Help},
+		{Ref: 2, Type: record.GetMetricType(m2.Type), Unit: m2.Unit, Help: m2.Help},
+		{Ref: 3, Type: record.GetMetricType(m3.Type), Unit: m3.Unit, Help: m3.Help},
+		{Ref: 4, Type: record.GetMetricType(m4.Type), Unit: m4.Unit, Help: m4.Help},
+		{Ref: 2, Type: record.GetMetricType(m5.Type), Unit: m5.Unit, Help: m5.Help},
+	}
+	require.Len(t, gotMetadataBlocks, 2)
+	require.Equal(t, expectedMetadata[:3], gotMetadataBlocks[0])
+	require.Equal(t, expectedMetadata[3:], gotMetadataBlocks[1])
+}
+
+func TestMetadataCheckpointingOnlyKeepsLatestEntry(t *testing.T) {
+	ctx := context.Background()
+	numSamples := 10000
+	hb, w := newTestHead(t, int64(numSamples)*10, false)
+
+	// Add some series so we can append metadata to them.
+	app := hb.Appender(ctx)
+	s1 := labels.FromStrings("a", "b")
+	s2 := labels.FromStrings("c", "d")
+	s3 := labels.FromStrings("e", "f")
+	app.Append(0, s1, 0, 0)
+	app.Append(0, s2, 0, 0)
+	app.Append(0, s3, 0, 0)
+	app.Commit()
+
+	// Add a first round of metadata to the first three series.
+	// Re-take the Appender, as the previous Commit will have it closed.
+	app = hb.Appender(ctx)
+	m1 := metadata.Metadata{Type: "gauge", Unit: "unit_1", Help: "help_1"}
+	m2 := metadata.Metadata{Type: "gauge", Unit: "unit_2", Help: "help_2"}
+	m3 := metadata.Metadata{Type: "gauge", Unit: "unit_3", Help: "help_3"}
+	app.AppendMetadata(0, s1, m1)
+	app.AppendMetadata(0, s2, m2)
+	app.AppendMetadata(0, s3, m3)
+
+	// Update metadata for first series.
+	m4 := metadata.Metadata{Type: "counter", Unit: "unit_4", Help: "help_4"}
+	app.AppendMetadata(0, s1, m4)
+
+	// Switch back-and-forth metadata for second series.
+	// Since it ended on the same metadata record, we don't expect a new
+	// addition here.
+	m5 := metadata.Metadata{Type: "counter", Unit: "unit_5", Help: "help_5"}
+	app.AppendMetadata(0, s2, m5)
+	app.AppendMetadata(0, s2, m2)
+	app.AppendMetadata(0, s2, m5)
+	app.AppendMetadata(0, s2, m2)
+	app.AppendMetadata(0, s2, m5)
+	app.AppendMetadata(0, s2, m2)
+
+	app.Commit()
+
+	// Let's cause a checkpoint.
+	first, last, err := wal.Segments(w.Dir())
+	require.NoError(t, err)
+	_, err = wal.Checkpoint(log.NewNopLogger(), w, first, last-1, func(id chunks.HeadSeriesRef) bool { return true }, 0)
```

Comment:
We can have `keep` to return false for one of the series to test if it excludes the deleted series. Maybe have another simple s4-m4 and return false for s3 (ref=3).